relay-filter/src/config.rs

```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
     }
 }
 
+/// Configuration of generic filters combined from project and global config.
+///
+/// See [`DynamicFiltersConfigIter`] for details on how to iterate easily
+/// through the applicable filters.
+#[derive(Copy, Clone, Debug)]
+pub(crate) struct DynamicFiltersConfig<'a> {
+    /// Configuration of generic filters from project configs.
+    project_filters: &'a GenericFiltersConfig,
+    /// Configuration of generic filters from global config.
+    global_filters: Option<&'a GenericFiltersConfig>,
+    /// Maximum supported version for generic filters.
+    ///
+    /// It applies to all filters, from both project and global configs.
+    max_version: u16,
+}
+
+impl<'a> DynamicFiltersConfig<'a> {
+    /// Creates a [`DynamicFiltersConfig`] from the project and global configs.
+    pub fn new(
+        project_filters: &'a GenericFiltersConfig,
+        global_filters: Option<&'a GenericFiltersConfig>,
+        max_version: u16,
+    ) -> Self {
+        DynamicFiltersConfig {
+            project_filters,
+            global_filters,
+            max_version,
+        }
+    }
+}
+
+impl<'a> IntoIterator for DynamicFiltersConfig<'a> {
+    type Item = &'a GenericFilterConfig;
```

Comment:
I believe we can simplify the iterator logic by a lot if we had an `Item` type `MergedGenericFilterConfig<'a>` here, that offers methods to retrieve the merged config. Then the iterator would be something like

```rust
match project_filters.next() {
    Some(pf) => return MergedConfig::new(pf, global_filters.get(pf.id)),
    None => match global_filters.next() {
        Some(gf) => return MergedConfig::new(gf, None),
        None => None,
    }
}
```

Reply:
Assuming the merging logic happens in `MergedGenericFilterConfig` or `MergedConfig`, how would the indexes for the next filter be updated? IIUC this can't happen in the `Item` of the iterator.

On the other hand, I don't like creating an abstraction that returns an abstraction over filters. As the client of the API, I just want to easily get the filters back in an easy single step.

Reply:
You wouldn't even need to track any indexes, the code above would already give you the body for the `next()` implementation.

Since there's a lot of comments on this PR, what do you think about a live discussion / some pair programming tomorrow?

Reply:
Followed the simplified approach in [a6db146](https://github.com/getsentry/relay/pull/3161/commits/a6db146fd2cbf1c653f989bbb05790e425e8023d).